### ISSUE TITLE
 checkbox group

### ISSUE THREAD

**Issue by mbeckenbach** [angular/material2#4926]
#### Bug, feature request, or proposal:
Feature request

#### What is the expected behavior?
Similar to the multiple select, it should be possible to have a list of check boxes to select multiple options returned in a single array of choices. A required validator would check for at least one selection.

#### What is the current behavior?
Has to be made via select or custom logic.

#### What are the steps to reproduce?

Providing a Plunker (or similar) is the *best* way to get the team to see your issue.
Plunker template: https://goo.gl/DlHd6U

#### What is the use-case or motivation for changing an existing behavior?
Need it quite often for my projects. Sometimes a select drop down is not what customers wish for a multiple selection. Sometimes it is better to see all options without having to open a menu.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Ng4.1.x, MD beta 6

**Comment by crisbeto** [comment]
I believe that you're looking for a radio group, which is supported already.

**Comment by jelbourn** [comment]
Not sure this is common enough of a pattern to warrant including in Angular Material

**Comment by lobo-tuerto** [comment]
I think this is common enough, I have had to create some custom checkbox-list inputs before.

@crisbeto The difference with the radio group is that you can't have more than 1 selected item, this covers it.

**Comment by gseabrook** [comment]
I'd agree this is a pretty common use case, particularly once the data table is ready, as selecting items from a table happens all the time. Its also on the demo page for Material 1 under the syncing section (https://material.angularjs.org/latest/demo/checkbox)

**Comment by Toub** [comment]
This is a common use case, exactly as radio and select ones.
Related to https://github.com/angular/angular/issues/10948

**Comment by deanchalk** [comment]
I would like to build this and add it to angular material as a community contribution. is this possible (I've not contributed to a community project before - so forgive my ignorance). I have read and understood the community contribution guidelines

**Comment by jelbourn** [comment]
@deanchalk what do you imagine this would look like?

**Comment by deanchalk** [comment]
Im imagining a checkbox container component that exposes a value property and valuechanged event. The child checkbox components could have a custom directive to allow a value to be associated with each one, and then the checkbox container will emit the valuechanged event whenever any of the checkboxes change their checked state and the value will be an array of the value's of all of the checked checkboxes.

This kind of component will be useful in forms where you want to capture a value that is an array of values (like a list of choices etc)

**Comment by jelbourn** [comment]
How is that different from the existing `mat-selection-list`?

**Comment by deanchalk** [comment]
you're right, I hadn't thought about it that way. I'll find another issue as a possible way of making my first contribution - thanks

**Comment by xagon** [comment]
Bumping this one up. Much desired.

**Comment by Dimasik11** [comment]
bumping.  i i thibk is desired.

**Comment by Londovir** [comment]
@jelbourn mat-selection-list (at least, as demonstrated on the Material 2 components list) seems to have a few issues that make it dissimilar to what is discussed here.

First, the appearance of the checkbox on the demo is automatically placed on the far right edge of the list item. This doesn't appear to agree with the most recent Google Material Design spec, specifically [Lists: Controls](https://material.io/guidelines/components/lists-controls.html#lists-controls-usage), which show checkboxes as a primary action/state on the left edge.

Second, it appears that mat-selection-list does not have defined value accessors, and therefore does not support ngModel or reactive forms. This makes it more unsuitable/"unfriendly" for use in forms, which is certainly a defined use-case of multiple selection checkbox units per [Selection controls](https://material.io/guidelines/components/selection-controls.html).

**Comment by crisbeto** [comment]
[The selection list definitely has a value accessor](https://github.com/angular/material2/blob/master/src/lib/list/selection-list.ts#L288) @Londovir, have you run into any particular issues with it?

**Comment by Londovir** [comment]
@crisbeto My apologies - I was looking at the API documentation on the Angular Material website, along with the example mat-selection-list shown there (the one with the shoe types). I then forked the stackblitz example from there and attempted to add both [(ngModel)] and [formGroup] style binding, at it would not work. I must assume that is because the stackblitz used on the official documentation site is using 5.0.0-rc.3 (took me a while to notice that), and ngModel binding wasn't supported until the full 5.0.0 release. Looks like the API docs could use a little refreshing, and the stackblitz could use an upgrade to 5.0.0 as well and all would be well.

**Comment by danwellman** [comment]
The problem is that it doesn't appear that the select list can be validated as a group - if you assign a custom validator to a select list, the validator will only be invoked with individual checkboxes within the list, when an individual list item loses focus. This is confusing and inconsistent with other Material controls. 

What would be more useful is if the validator was invoked for the list of checkboxes whenever the checked state of any checkboxes within the list changes, when the list as a whole loses focus.

**Comment by duayres** [comment]
Bumping as it seens desired for me and my team too.

**Comment by jelbourn** [comment]
"Bumping" is not a thing. It only spams the maintainers.

**Comment by duayres** [comment]
My apologies, removing the posts

**Comment by VamsiDharmaji** [comment]
mat-selection-list with checkbox is not ADA friendly

**Comment by ravidannani** [comment]
Hi, Do we have an checkbox-group similar to radio-button-group? Please let me know.

**Comment by tommueller** [comment]
@ravidannani have a look at the [selection lists](https://material.angular.io/components/list/overview#selection-lists)

**Comment by ravidannani** [comment]
@tommueller Thanks for your reply. I actually need some kind of checkbox-group only (should be similar to radio-button-group). not selection lists. do you have a way to do that?

**Comment by VenkatakrishnaNelapati** [comment]
@ravidannani  check this StackOverflow question: https://stackoverflow.com/questions/40927167/angular-2-reactive-forms-array-of-checkbox-values

**Comment by sh977218** [comment]
I agree there should be Check Box List feature implemented. There is no current feature supports now.
Radio Button does not support check box list. Check box list supposed to be multiple selection.
Selection List does not support check box list Neither. Because select does not show the option all the time.

**Comment by clundberg1** [comment]
This feature definitely needs to be implemented. Even Material for Angularjs has it. 

A selection list is simply not enough because you shouldn't be restricted to showing multiple checkboxes in a list. What if I want to render them in a row like radio buttons? Not to mention you can't even pick which side the checkbox should appear on.

**Comment by Worthy7** [comment]
Has anyone considered simply adding an option to the mat-selection which shows it inline and expended instead?

**Comment by danieldanielecki** [comment]
For my case this worked really fine: [https://stackblitz.com/angular/eblgorgoedn?file=app%2Flist-selection-example.ts](https://stackblitz.com/angular/eblgorgoedn?file=app%2Flist-selection-example.ts)

**Comment by tommueller** [comment]
@danieldanielecki  The selection list works fine, but is not a proper Angular Form Element, so you cannot use it in Angular Forms ootb.

**Comment by danieldanielecki** [comment]
"ootb"? 

Was able to integrate it quite easily using reactive forms into contact form.

**Comment by nzbin** [comment]
@crisbeto @jelbourn @deanchalk @Londovir @danwellman @Chembeti @christianlundberg @tommueller 
There already has checkbox-group component in Material Extensions library. You can have a try.
https://ng-matero.github.io/extensions/components/checkbox-group/overview
